[sw] Clang-Tidy produces false positives

### Description

See [clang_tidy_results.txt](https://github.com/lowRISC/opentitan/files/12316798/clang_tidy_results.txt), which is the output of  `./bazelisk.sh test //quality:clang_tidy_check` at a76371a6c309727f0451a3e1cad5011410ef3dfb.

Here's a cherry-picked error about the use of an uninitialized value:

```
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/1269/execroot/lowrisc_opentitan/sw/device/silicon_creator/rom/rom.c:234:9: error: 1st function call argument is an uninitialized value [clang-analyzer-core.CallAndMessage,-warnings-as-errors]
        sigverify_spx_key_id_get(&ext_spx_key->key), lc_state, &spx_key));
        ^
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/1269/execroot/lowrisc_opentitan/sw/device/silicon_creator/rom/rom.c:532:3: note: Loop condition is false.  Exiting loop
  CFI_FUNC_COUNTER_INIT(rom_counters, kCfiRomMain);
  ^
./sw/device/silicon_creator/lib/cfi.h:118:3: note: expanded from macro 'CFI_FUNC_COUNTER_INIT'
  do {                                            \
  ^
/home/dan_zerorisc_com/.cache/bazel/_bazel_dan/e8b9a1cec6d6126ab719828a176ad974/sandbox/linux-sandbox/1269/execroot/lowrisc_opentitan/sw/device/silicon_creator/rom/rom.c:534:3: note: Loop condition is false.  Exiting loop

[I'm eliding a few hundred lines of static analyzer reasoning...]
```

The gist of it is that Clang-Tidy flagged [rom.c:234](https://cs.opensource.google/opentitan/opentitan/+/a76371a6c309727f0451a3e1cad5011410ef3dfb:sw/device/silicon_creator/rom/rom.c;l=234) because it thinks `&ext_spx_key->key` could be undefined.


```c
 const sigverify_spx_key_t *spx_key;
  const sigverify_spx_signature_t *spx_signature;
  uint32_t sigverify_spx_en = sigverify_spx_verify_enabled(lc_state);
  if (launder32(sigverify_spx_en) != kSigverifySpxDisabledOtp) {
    const manifest_ext_spx_key_t *ext_spx_key;
    HARDENED_RETURN_IF_ERROR(manifest_ext_get_spx_key(manifest, &ext_spx_key));
    HARDENED_RETURN_IF_ERROR(sigverify_spx_key_get(
        sigverify_spx_key_id_get(&ext_spx_key->key), lc_state, &spx_key));
```


I think the issue was that it couldn't see through `HARDENED_RETURN_IF_ERROR` because it's implemented with `launder32()`.

@luismarques I wrote up a more minimal input that reproduces the uninitialized value warning. Do you think there's a feasible way to improve Clang-Analyzer to "see through" the inline assembly? 

For convenience, here's the repro case on Compiler Explorer with a clang-tidy tab on the right: https://godbolt.org/z/Kbd3P853c

And for future proofing, here's the code. I'll have to play with this a little more tomorrow.

```c
// This file is a minimal test case for reproducing the false-positive
// Clang-Analyzer results documented in issue #19412.
//
// https://github.com/lowRISC/opentitan/issues/19412.

typedef int err_t;
const err_t kError = 0xff;
const err_t kOk = 0;

typedef struct foo {
  int data;
} foo_t;

foo_t global_foo = {.data = 42};
int global_decision = 0;

// This function is based on OpenTitan's `launder32()`.
int launder(int val) {
  asm volatile("" : "+r"(val));
  return val;
}

// This macro aborts the calling function when `expr_` is not `kOk`.
#define HARDENED_RETURN_IF_ERROR(expr_) \
  do {                                  \
    err_t err_ = expr_;                 \
    if (launder(err_) != kOk) {         \
      return err_;                      \
    }                                   \
  } while (0)

#define SOFTENED_RETURN_IF_ERROR(expr_) \
  do {                                  \
    err_t err_ = expr_;                 \
    if (err_ != kOk) {                  \
      return err_;                      \
    }                                   \
  } while (0)

err_t process(int val);

err_t maybe_initialize_foo(const struct foo **foo) {
  if (global_decision) {
    *foo = &global_foo;
    return kOk;
  }
  return kError;
}

// This function reproduces the false positive from Clang-Analyzer documented in
// issue #19412: https://github.com/lowRISC/opentitan/issues/19412.
//
// [...]sw/device/lib/base/minimal.c:58:36: error: Access to field 'data' results in a dereference of an undefined pointer value (loaded from variable 'foo') [clang-analyzer-core.NullDereference,-warnings-as-errors]
//   HARDENED_RETURN_IF_ERROR(process(foo->data));
//                                    ^
err_t repro(void) {
  const struct foo *foo;
  HARDENED_RETURN_IF_ERROR(maybe_initialize_foo(&foo));
  // If control reaches this point, `foo` is initialized.
  HARDENED_RETURN_IF_ERROR(process(foo->data));
  return kOk;
}

// This function does not reproduce the false positive.
err_t no_repro(void) {
  const struct foo *foo;
  SOFTENED_RETURN_IF_ERROR(maybe_initialize_foo(&foo));
  // If control reaches this point, `foo` is initialized.
  SOFTENED_RETURN_IF_ERROR(process(foo->data));
  return kOk;
}
```

--- a/rules/quality.bzl
+++ b/rules/quality.bzl
@@ -195,6 +195,8 @@ def _cc_aspect_impl(target, ctx, action_callback):
 #  ./bazelisk.sh run @lowrisc_rv32imcb_files//:bin/clang-tidy -- --checks='*' --list-checks
 _CLANG_TIDY_CHECKS = [
     "clang-analyzer-core.*",
+    # Disable advice to replace `memcpy` with `mempcy_s`.
+    "-clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling",
 ]
 
 def _clang_tidy_run_action(ctx, cc_toolchain, cc_compile_ctx, generated_file, command_line, src):

--- a/sw/device/lib/base/hardened.h
+++ b/sw/device/lib/base/hardened.h
@@ -240,6 +240,12 @@ inline uint32_t launder32(uint32_t val) {
   // > instead require that reordering be prevented through careful sequencing
   // > of statements.
 
+  // When we're building for static analysis, reduce false positives by
+  // short-circuiting the inline assembly block.
+#if OT_BUILD_FOR_STATIC_ANALYZER
+  return val;
+#endif
+
   // The +r constraint tells the compiler that this is an "inout" parameter: it
   // means that not only does the black box depend on `val`, but it also mutates
   // it in an unspecified way.

--- a/sw/device/lib/base/macros.h
+++ b/sw/device/lib/base/macros.h
@@ -453,6 +453,17 @@ extern "C++" {
 #define OT_USED __attribute__((used))
 
 /**
+ * OT_BUILD_FOR_STATIC_ANALYZER indicates whether we are compiling for the
+ * purpose of static analysis. Currently, this macro only detects
+ * Clang-Analyzer, which is used as a backend by Clang-Tidy.
+ */
+#ifdef __clang_analyzer__
+#define OT_BUILD_FOR_STATIC_ANALYZER 1
+#else
+#define OT_BUILD_FOR_STATIC_ANALYZER 0
+#endif
+
+/**
  *  This macro is used to align an offset to point to a 32b value.
  */
 #define OT_ALIGN_MEM(x) (uint32_t)(4 + (((uintptr_t)(x)-1) & ~3u))

--- a/sw/device/lib/base/memory.c
+++ b/sw/device/lib/base/memory.c
@@ -4,6 +4,7 @@
 
 #include "sw/device/lib/base/memory.h"
 
+#include <assert.h>
 #include <stdint.h>
 
 #include "sw/device/lib/base/macros.h"
@@ -121,6 +122,10 @@ int OT_PREFIX_IF_NOT_RV32(memcmp)(const void *lhs, const void *rhs,
     }
   }
   for (; i < tail_offset; i += sizeof(uint32_t)) {
+#if OT_BUILD_FOR_STATIC_ANALYZER
+    assert(&lhs8[i] != NULL);
+    assert(&rhs8[i] != NULL);
+#endif
     uint32_t word_left = __builtin_bswap32(read_32(&lhs8[i]));
     uint32_t word_right = __builtin_bswap32(read_32(&rhs8[i]));
     static_assert(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__,
@@ -157,6 +162,10 @@ int memrcmp(const void *lhs, const void *rhs, size_t len) {
   }
   for (; end > body_offset; end -= sizeof(uint32_t)) {
     const size_t i = end - sizeof(uint32_t);
+#if OT_BUILD_FOR_STATIC_ANALYZER
+    assert(&lhs8[i] != NULL);
+    assert(&rhs8[i] != NULL);
+#endif
     uint32_t word_left = read_32(&lhs8[i]);
     uint32_t word_right = read_32(&rhs8[i]);
     static_assert(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__,

--- a/sw/device/silicon_creator/rom/rom.c
+++ b/sw/device/silicon_creator/rom/rom.c
@@ -4,6 +4,7 @@
 
 #include "sw/device/silicon_creator/rom/rom.h"
 
+#include <assert.h>
 #include <stdbool.h>
 #include <stdint.h>
 
@@ -224,8 +225,8 @@ static rom_error_t rom_verify(const manifest_t *manifest,
   HARDENED_RETURN_IF_ERROR(sigverify_rsa_key_get(
       sigverify_rsa_key_id_get(&manifest->rsa_modulus), lc_state, &rsa_key));
 
-  const sigverify_spx_key_t *spx_key;
-  const sigverify_spx_signature_t *spx_signature;
+  const sigverify_spx_key_t *spx_key = NULL;
+  const sigverify_spx_signature_t *spx_signature = NULL;
   uint32_t sigverify_spx_en = sigverify_spx_verify_enabled(lc_state);
   if (launder32(sigverify_spx_en) != kSigverifySpxDisabledOtp) {
     const manifest_ext_spx_key_t *ext_spx_key;
@@ -474,6 +475,10 @@ static rom_error_t rom_boot(const manifest_t *manifest, uint32_t flash_exec) {
       HARDENED_TRAP();
   }
   HARDENED_CHECK_EQ(manifest, manifest_check);
+
+#if OT_BUILD_FOR_STATIC_ANALYZER
+  assert(manifest_check != NULL);
+#endif
 
   if (launder32(manifest_check->address_translation) == kHardenedBoolTrue) {
     HARDENED_CHECK_EQ(manifest_check->address_translation, kHardenedBoolTrue);

